File: internal/httpapi/partial_date_test.go
package httpapi

import "testing"

// normalizePartialDate backs every partial date in the app: the shelf's read log
// (started/finished) and, from §24, a quote's occasion. Its own comment promises
// it rejects "the shapes the regexp lets through but a calendar would not, so a
// stored date is always a real one" — and it used to check only that the day was
// between 1 and 31, which accepted 30 February and 31 April.
func TestNormalizePartialDate(t *testing.T) {
	cases := []struct {
		name  string
		in    string
		valid bool
	}{
		// The three shapes it accepts, plus the leap day and the empty value.
		{"a bare year", "1944", true},
		{"a year and a month", "1944-01", true},
		{"a full date", "1944-01-23", true},
		{"the 2024 leap day", "2024-02-29", true},
		{"an empty value", "", true},

		// The whole point of a partial date is that it can be vague. It cannot be
		// wrong: "1944" is honest, "1944-02-30" is a typo that would sort between
		// two real dates and never resolve to a day.
		{"a date that does not exist: 30 February", "1944-02-30", false},         // February has never had 30 days
		{"a date that does not exist: 31 April", "1944-04-31", false},            // nor April 31
		{"a date that does not exist: 29 February in 2023", "2023-02-29", false}, // 2023 is not a leap year
		{"a date that does not exist: 29 February in 1900", "1900-02-29", false}, // nor 1900 — divisible by 100, not by 400

		// Malformed input: the shapes that are not dates at all.
		{"malformed: a two-digit year", "44", false},
		{"malformed: prose", "not a date", false},
		{"malformed: month 13", "1944-13", false},
		{"malformed: month 00", "1944-00", false},
		{"malformed: day 32", "1944-01-32", false},
		{"malformed: day 00", "1944-01-00", false},
		{"malformed: a year below the accepted range", "0999", false},
		{"malformed: a year above the accepted range", "3001", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			v := tc.in
			msg := normalizePartialDate("date", &v)
			if tc.valid && msg != "" {
				t.Errorf("normalizePartialDate(%q) rejected it: %s", tc.in, msg)
			}
			if !tc.valid && msg == "" {
				t.Errorf("normalizePartialDate(%q) accepted it", tc.in)
			}
		})
	}
}

// The message names the field, because the same validator serves several and
// "must be YYYY..." alone would not say which one to fix.
func TestPartialDateNamesTheField(t *testing.T) {
	v := "nonsense"
	msg := normalizePartialDate("occasion date", &v)
	if msg == "" || msg[:len("occasion date")] != "occasion date" {
		t.Fatalf("message does not lead with the field name: %q", msg)
	}
}

func TestPartialDateTrims(t *testing.T) {
	v := "  1944-01-23  "
	if msg := normalizePartialDate("date", &v); msg != "" {
		t.Fatalf("rejected a padded date: %s", msg)
	}
	if v != "1944-01-23" {
		t.Fatalf("did not trim in place: %q", v)
	}
}

File: web/frontend/test/pure/partial-date.test.js
// The partial-date helpers (ui.jsx): the client mirror of the server's
// normalizePartialDate.
//
// They had no test at all, which is worth stating plainly because the server
// half was found to be WRONG on exactly this ground: it checked only that the
// day was between 1 and 31, so it accepted 30 February and 31 April and stored
// them. That was fixed by running the full-date case through a real calendar.
// This file pins the same rule on the client, where the check exists so a typo
// is caught before it is sent rather than bounced back as a 400.
//
// The three shapes are the point of the format. A partial date is allowed to be
// VAGUE — "1944" is an honest answer about a speech — but it is not allowed to
// be WRONG, because these sort against each other as text and a date that never
// happened would file itself neatly between two that did.

import { describe, expect, it } from 'vitest'
import { formatPartialDate, isPartialDate, todayPartial } from '../../src/ui.jsx'

describe('isPartialDate', () => {
  it('accepts the three shapes', () => {
    for (const good of ['1944', '1944-01', '1944-01-23', '2024-02-29']) {
      expect(isPartialDate(good), good).toBe(true)
    }
  })

  // One test over all fifteen rejected inputs rather than three loops: every
  // one is the same assertion — isPartialDate is strictly false — and only the
  // string differs. The three groups keep their headings as comments, and the
  // aggregate names every input that was wrongly accepted rather than stopping
  // at the first.
  it('rejects a date that is malformed, out of range, or not on the calendar', () => {
    const bad = [
      // rejects a date the calendar does not have
      '1944-02-30', // February has never had 30 days
      '1944-04-31', // nor April 31
      '2023-02-29', // 2023 is not a leap year
      '1900-02-29', // nor 1900 — divisible by 100, not by 400
      // rejects a month outside the year
      '1944-00',
      '1944-13',
      '1944-01-00',
      '1944-01-32',
      // rejects anything that is not the shape
      '',
      '44',
      '1944-1',
      '1944-01-2',
      'not a date',
      '1944/01/23',
      '1944-01-23T00:00:00',
    ]
    const accepted = bad.filter((s) => isPartialDate(s) !== false).map((s) => JSON.stringify(s))
    expect(accepted).toEqual([])
  })

  it('rejects a year outside the bounds it claims', () => {
    expect(isPartialDate('0999')).toBe(false)
    expect(isPartialDate('3001')).toBe(false)
    expect(isPartialDate('1000')).toBe(true)
    expect(isPartialDate('3000')).toBe(true)
  })
})

describe('formatPartialDate', () => {
  // The precision SHOWS, which is the whole reason for keeping a partial date
  // rather than padding it to a day nobody recorded.
  it('renders each shape at its own precision', () => {
    expect(formatPartialDate('1944')).toBe('1944')
    expect(formatPartialDate('1944-03')).toBe('Mar 1944')
    expect(formatPartialDate('1944-03-04')).toBe('4 Mar 1944')
  })

  it('renders nothing for nothing', () => {
    expect(formatPartialDate('')).toBe('')
    expect(formatPartialDate(null)).toBe('')
    expect(formatPartialDate(undefined)).toBe('')
  })

  // A leading zero in the day is dropped, because "04 Mar" reads as a form
  // field and "4 Mar" reads as a date.
  it('does not pad the day', () => {
    expect(formatPartialDate('1944-03-04')).toBe('4 Mar 1944')
    expect(formatPartialDate('1944-12-25')).toBe('25 Dec 1944')
  })

  // Every month index has to land on the right name — an off-by-one here is
  // invisible in January and wrong for the other eleven.
  it('names all twelve months', () => {
    const want = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
    want.forEach((name, i) => {
      const mm = String(i + 1).padStart(2, '0')
      expect(formatPartialDate(`1944-${mm}`), mm).toBe(`${name} 1944`)
    })
  })
})

describe('todayPartial', () => {
  it('is a full date the validator accepts', () => {
    expect(isPartialDate(todayPartial())).toBe(true)
    expect(todayPartial()).toMatch(/^\d{4}-\d{2}-\d{2}$/)
  })

  // It builds the string from LOCAL parts rather than toISOString, which would
  // hand back yesterday's date for anyone west of UTC after their evening.
  it('agrees with the local calendar, not UTC', () => {
    const n = new Date()
    const p = (x) => String(x).padStart(2, '0')
    expect(todayPartial()).toBe(`${n.getFullYear()}-${p(n.getMonth() + 1)}-${p(n.getDate())}`)
  })
})
